lib/srv/usermgmt.go

```diff
@@ -0,0 +1,297 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package srv
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os/user"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/services/local"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"github.com/siddontang/go/log"
+)
+
+// NewHostUsers initialize a new HostUsers object
+func NewHostUsers(ctx context.Context, storage *local.PresenceService) (HostUsers, error) {
+	backend, err := newHostUsersBackend()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cancelCtx, cancelFunc := context.WithCancel(ctx)
+	return &HostUserManagement{
+		backend: backend,
+		ctx:     cancelCtx,
+		cancel:  cancelFunc,
+		storage: storage,
+	}, nil
+}
+
+type HostUsersBackend interface {
+	// GetAllUsers returns all host users on a node.
+	GetAllUsers() ([]string, error)
+	// UserGIDs returns a list of group ids for a user.
+	UserGIDs(*user.User) ([]string, error)
+	// Lookup retrieves a user by name.
+	Lookup(name string) (*user.User, error)
+	// LookupGroup retrieves a group by name.
+	LookupGroup(group string) (*user.Group, error)
+	// CreateGroup creates a group on a host.
+	CreateGroup(group string) error
+	// CreateUser creates a user on a host.
+	CreateUser(name string, groups []string) error
+	// DeleteUser deletes a user from a host.
+	DeleteUser(name string) error
+}
+
+// HostUsersProvisioningBackend is used to implement HostUsersBackend
+type HostUsersProvisioningBackend struct{}
+
+type userCloser struct {
+	users    HostUsers
+	backend  HostUsersBackend
+	username string
+}
+
+func (u *userCloser) Close() error {
```

Comment:
This should also grab the host user management lock and check for expiration - again, what if a session ends right as one is halfway through starting?